File: C++/strings/z-function.cpp
#include <iostream>
#include <string>
#include <vector>

std::vector<int> zFunction(std::string line){
    int sz = line.size();
    std::vector<int> result(sz);
    result[0] = 0;
    std::pair<int, int> z_block{0, 0};
    for (int index = 1; index < sz; index++){
        if (index <= z_block.second){
            int now = std::min(result.at(index - z_block.first), z_block.second - index + 1);
            if (now == 0){
                result.at(index) = 0;
                continue;
            }
            int next_1 = now;
            int next_2 = z_block.second + 1;
            while (next_2 < sz and line.at(next_1) == line.at(next_2)){
                next_1++;
                next_2++;
            }
            if (next_2 - 1  > z_block.second){ 
                z_block.second = next_2 - 1;
                z_block.first = index;
            }
            result.at(index) = next_1;
        }
        else{
            int i = 0;
            int j = index;
            while (j < sz and  line.at(i) == line.at(j)){
                i++;
                j++;
            }
            if (i != 0){
                z_block.first = index;
                z_block.second = j - 1;
            }
            result.at(index) = i;
        }
    }
    return result;
}

int main(){

    std::string line;
    std::getline(std::cin, line);
    std::vector<int> answer = zFunction(line);
    for (int index = 0; index < line.size(); index++){
        printf("%d ", answer.at(index));
    }
    
    return 0;
}

File: Python/strings/z-function.py
def z_function(string: str) -> list[int]:
    sz = len(string)
    zblock = [0, 0]
    result = [0] * sz
    for index in range(1, sz):
        if index <= zblock[1]:
            result[index] = min(result[index - zblock[0]], zblock[1] - index + 1)

        left = result[index]
        while index +  left < sz and string[left] == string[left + index]:
            left += 1

        if left > zblock[1]:
            zblock[1] = left - 1 + index
            zblock[0] = index

        result[index] = left

    return result

s = input()
print(*z_function(s))

